### ISSUE TITLE
adding custom %TAG entries to file header

### ISSUE THREAD

**Issue by CagtayFabry** [asdf-format/asdf#890]
I think it would be nice to provide a way for custom extensions to add additional `%TAG` directives to the top YAML header to shorten tags provided by custom extensions that might lead to long tag strings. ASDF already does this by introducing `%TAG ! tag:stsci.edu:asdf/` .

Consider the following example:
```yaml
#ASDF 1.5.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.1.0
# asdf headers
fraction: !<tag:nowhere.org:custom/fraction-1.0.0> [10, 3]
...
```
With adding  `%TAG !foo! tag:nowhere.org:custom/` we get the shorter
```yaml
#ASDF 1.5.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
%TAG !foo! tag:nowhere.org:custom/
--- !core/asdf-1.1.0
# asdf headers
fraction: !foo!fraction-1.0.0 [10, 3]
...
```

The only thing needed to get this output is to add the `"!foo!": "tag:nowhere.org:custom/"` to the `tags` mapping in the final `    yaml.dump_all` call in `asdf.yamlutil.dump_tree` here:
https://github.com/asdf-format/asdf/blob/8ccee8b17a053d6e33ebca81a4ed3f3c3569aee1/asdf/yamlutil.py#L408-L413

Currently `tags` is (statically) build to simply consist of `{"!": "tag:stsci.edu:asdf/"}`

One possible way to implement this would be to introduce a new (optional) property for the ExtensionClass that provides the tags.
Example extending the `FractionExtension` with a new property `yaml_tag_context`:
```python
class FractionExtension(asdf.AsdfExtension):
    # ...

    @property
    def yaml_tag_context(self):
        return {"!foo!": "tag:nowhere.org:custom/"}
```
We can now query this property for registered extensions and add them to the list.

I tested this for simple read/write and it works, see this simple diff  https://github.com/asdf-format/asdf/compare/master...CagtayFabry:test_custom_yaml_tag

~Of course it would be nice to only load `%TAG` values from extensions that are actually used when creating the file (similar to listing them in the history)~ update

**Comment by eslavich** [comment]
Hey cool I didn't know that was possible in YAML!  I like this idea.  We already [track](https://github.com/asdf-format/asdf/blob/master/asdf/yamlutil.py#L250) which extensions were used when writing the file so it should be straightforward to only add the necessary `%TAG` directives.

**Comment by CagtayFabry** [comment]
> Hey cool I didn't know that was possible in YAML! I like this idea. We already [track](https://github.com/asdf-format/asdf/blob/master/asdf/yamlutil.py#L250) which extensions were used when writing the file so it should be straightforward to only add the necessary `%TAG` directives.

yes, getting the extensions from `_serialization_context` makes this really simple 👍 
I updated the example above https://github.com/asdf-format/asdf/compare/master...CagtayFabry:test_custom_yaml_tag

**Comment by perrygreenfield** [comment]
Yes, I was also under the impression that only one of these %TAG was permitted. That would be a nice feature. I've been wondering how we could make tags abbreviated.

**Comment by CagtayFabry** [comment]
I am not specifically familiar with YAML specs in this regard but it appears multiple (different) handles are allowed in YAML 1.1 and 1.2

from https://yaml.org/spec/1.1/

> ### Example 7.6. Tag Prefixes
> ```yaml
> %TAG !      !foo
> %TAG !yaml! tag:yaml.org,2002:
> ---
> - !bar "baz"
> - !yaml!str "string"
> ```

also, multiple directives must not define the same prefix (I assume multiple handles in principle are allowed)
> It is an error to specify more than one “TAG” directive for the same handle in the same document, even if both occurrences give the same prefix.

**Comment by eslavich** [comment]
Resolved by #963
